pkg/kubelet/deviceplugin/manager.go

```diff
@@ -0,0 +1,206 @@
+	removeContents(m.socketdir)
+
+	if err := os.Remove(socketPath); err != nil && !os.IsNotExist(err) {
+		glog.Errorf(ErrRemoveSocket+" %+v", err)
+		return err
+	}
+
+	s, err := net.Listen("unix", socketPath)
+	if err != nil {
+		glog.Errorf(ErrListenSocket+" %+v", err)
+		return err
+	}
+
+	m.server = grpc.NewServer([]grpc.ServerOption{}...)
+
+	pluginapi.RegisterRegistrationServer(m.server, m)
+	go m.server.Serve(s)
+
+	return nil
+}
+
+// Devices is the map of devices that are known by the Device
+// Plugin manager with the kind of the devices as key
+func (m *ManagerImpl) Devices() map[string][]*pluginapi.Device {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	devs := make(map[string][]*pluginapi.Device)
+	for k, e := range m.Endpoints {
+		glog.V(3).Infof("Endpoint: %+v: %+v", k, e)
+		e.mutex.Lock()
+		devs[k] = copyDevices(e.devices)
+		e.mutex.Unlock()
+	}
+
+	return devs
+}
+
+// Allocate is the call that you can use to allocate a set of devices
+// from the registered device plugins.
+func (m *ManagerImpl) Allocate(resourceName string, devs []string) (*pluginapi.AllocateResponse, error) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	if len(devs) == 0 {
+		return nil, nil
+	}
+
+	glog.V(3).Infof("Recieved allocation request for devices %v for device plugin %s",
+		devs, resourceName)
+	e, ok := m.Endpoints[resourceName]
+	if !ok {
+		return nil, fmt.Errorf("Unknown Device Plugin %s", resourceName)
+	}
+
+	return e.allocate(devs)
+}
+
+// Register registers a device plugin.
+func (m *ManagerImpl) Register(ctx context.Context,
+	r *pluginapi.RegisterRequest) (*pluginapi.Empty, error) {
+	glog.V(2).Infof("Got request for Device Plugin %s", r.ResourceName)
+	if r.Version != pluginapi.Version {
+		return &pluginapi.Empty{},
+			fmt.Errorf(pluginapi.ErrUnsuportedVersion)
+	}
+
+	if err := IsResourceNameValid(r.ResourceName); err != nil {
+		return &pluginapi.Empty{}, err
+	}
+
+	// TODO: for now, always accepts newest device plugin. Later may consider to
+	// add some policies here, e.g., verify whether an old device plugin with the
+	// same resource name is still alive to determine whether we want to accept
+	// the new registration.
+	go m.addEndpoint(r)
+
+	return &pluginapi.Empty{}, nil
+}
+
+// Stop is the function that can stop the gRPC server.
+func (m *ManagerImpl) Stop() error {
+	for _, e := range m.Endpoints {
+		e.stop()
+	}
+	m.server.Stop()
+	return nil
+}
+
+func (m *ManagerImpl) addEndpoint(r *pluginapi.RegisterRequest) {
+	socketPath := filepath.Join(m.socketdir, r.Endpoint)
+	e, err := newEndpoint(socketPath, r.ResourceName, m.callback)
+	if err != nil {
+		glog.Errorf("Failed to dial device plugin with request %v: %v", r, err)
+		return
+	}
+
+	stream, err := e.list()
+	if err != nil {
+		glog.Errorf("Failed to List devices for plugin %v: %v", r.ResourceName, err)
+		return
+	}
+
+	go func() {
+		e.listAndWatch(stream)
+
+		m.mutex.Lock()
+		e.mutex.Lock()
```

Comment:
What are the invocation of endpoint object that are not already protected?
It would be ideal to at-least move the endpoint lock to the endpoint methods. For example, have a delete method in the endpoint object.

Reply:
This part is from Renaud's commits. Looking at again, I think we can actually remove the endpoint locking here and below. AFAIU, they don't seem to be needed.